core/src/broadcast_stage/broadcast_utils.rs

```diff
@@ -26,44 +28,85 @@ pub struct UnfinishedSlotInfo {
     pub parent: Slot,
 }
 
-/// This parameter tunes how many entries are received in one iteration of recv loop
-/// This will prevent broadcast stage from consuming more entries, that could have led
-/// to delays in shredding, and broadcasting shreds to peer validators
-const RECEIVE_ENTRY_COUNT_THRESHOLD: usize = 8;
+#[derive(Default)]
+pub(crate) struct RecvEntriesContext {
+    pending_entry: Option<WorkingBankEntry>,
+}
 
-pub(super) fn recv_slot_entries(receiver: &Receiver<WorkingBankEntry>) -> Result<ReceiveResults> {
+pub(super) fn recv_slot_entries(
+    ctx: &mut RecvEntriesContext,
+    receiver: &Receiver<WorkingBankEntry>,
+    receiver_coalesce_ms: u64,
+) -> Result<ReceiveResults> {
+    let target_batch_byte_count: u64 =
+        32 * ShredData::capacity(/*merkle_proof_size*/ None).unwrap() as u64;
+    let mut batch_byte_count: u64 = 8; // Vec len
     let timer = Duration::new(1, 0);
     let recv_start = Instant::now();
-    let (mut bank, (entry, mut last_tick_height)) = receiver.recv_timeout(timer)?;
+    let mut entries = Vec::default();
+    let mut pending_state = None;
+
+    if let Some((bank, (entry, last_tick_height))) = ctx.pending_entry.take() {
+        assert!(last_tick_height <= bank.max_tick_height());
+        batch_byte_count += serialized_size(&entry)?;
+        entries.push(entry);
+        pending_state = Some((bank, last_tick_height));
+    }
 
-    let mut entries = vec![entry];
+    let (mut bank, (entry, mut last_tick_height)) = match receiver.recv_timeout(timer) {
```

Comment:
I think ideally we want to:
* drain the channel as long as it has entries, and we have not hit `max_tick_height` yet.
* if we have not received enough entries to make _at least_ 32 shreds wait a bit longer to receive more entries.

So I think we would need something like:
```rust
// Wait for the very fisrt entry.
const TIMEOUT: Duration = Duration::from_secs(1);
let (mut bank, (entry, mut last_tick_height)) = receiver.recv_timeout(TIMEOUT)?;
                                                                                             
// Start the clock after receiving the first entry.
const COALESCE_DURATION: Duration = Duration::from_millis(100);
let mut recv_deadline = Instant::now() + COALESCE_DURATION;
                                                                                             
let mut entries = vec![entry];
                                                                                             
// Drain the channel as long as it has entries.
while last_tick_height != bank.max_tick_height() {
    let (try_bank, (entry, tick_height)) = match receiver.try_recv() {
        Err(_) => break,
        Ok(working_bank_entry) => working_bank_entry,
    };
    if try_bank.slot() != bank.slot() {
        recv_deadline = Instant::now() + COALESCE_DURATION;
        entries.clear();
        bank = try_bank;
    }
    last_tick_height = tick_height;
    entries.push(entry);
}
                                                                                             
// Wait for more entries to make at least 32 shreds.
let min_entries_size = 32 * ShredData::capacity(/*merkle_proof_size*/ None).unwrap() as u64;
let mut entries_size = bincode::serialized_size(&entries).unwrap();
while last_tick_height != bank.max_tick_height() && entries_size < min_entries_size {
    let (try_bank, (entry, tick_height)) = match receiver.recv_deadline(recv_deadline) {
        Err(_) => break,
        Ok(working_bank_entry) => working_bank_entry,
    };
    if try_bank.slot() != bank.slot() {
        recv_deadline = Instant::now() + COALESCE_DURATION;
        entries_size = 8; // Size of vector.
        entries.clear();
        bank = try_bank;
    }
    last_tick_height = tick_height;
    entries_size += bincode::serialized_size(&entry).unwrap();
    entries.push(entry);
}
```

Reply:
removed `ctx`, coalescing until data or time threshold is reached.